### ISSUE TITLE
Comparison to other approaches?

### ISSUE THREAD

**Issue by cscherrer** [thautwarm/CanonicalTraits.jl#7]
Hi Taine,

I was just talking to @McCoyBecker about traits. There are a few different approaches in Julia, and it can be tough to choose among them. He also mentioned being a fan of the approach Rust takes for this.

How would you compare these with `CanonicalTraits`?

**Comment by thautwarm** [comment]
`CanonicalTraits` is the Haskell way, which is very similar to that of Rust's, for both usage and implementation. I'd say the major difference between `CanonicalTraits` and traits/type classes in Haskell/Rust is, we cannot give constraints when defining a function. This is just a restriction of Julia(as a dynamic language).

**Comment by cscherrer** [comment]
Great, thanks! How about pros/cons vs Holy traits or SimpleTraits?

**Comment by thautwarm** [comment]
Sorry I found that I cannot give a short conclusion for this, and I have to have a rest because it's nearing the noon now...
Will reply this night.

**Comment by thautwarm** [comment]
This does deserve some detailed discussions.

**Comment by cscherrer** [comment]
Thanks, no rush. I do think it would be valuable for anyone considering options for traits :)

**Comment by McCoyBecker** [comment]
Correct me if I'm wrong, but can't you give type class constraints using the holy trick?

For example, `SimpleTraits.jl` provides this as part of an example:
```
@traitfn f(x::X) where {X; IsNice{X}} = "Very nice!"
@traitfn f(x::X) where {X; !IsNice{X}} = "Not so nice!"
```

Personally, I don't like that syntax - it seems clearer to write:
```
@traitfn f(x::X) where {IsNice{X}} ...
```
to remove boilerplate from the bound.

**Comment by thautwarm** [comment]
@McCoyBecker If the trait functions of `IsNice` is not statically checked when defining `@traitfn f(x::X) where {X; IsNice{X}}`, it still satisfys my previous description.

In `CanonicalTraits`, you can do this
```julia
@trait Add1{T} where
    add1 :: [X] => X
end
function f(x::X) where X
       Add1(X) # check if implemented
      "very nice"
end
```

**Comment by McCoyBecker** [comment]
Ah the difference is static and dynamic - got it! 

Is the below correct (?):

In Rust for example, a trait bound like the above is compiled away - at compile time, Rust creates methods for dispatch for all types which satisfy the bound. But here, the bound is checked dynamically and could only throw a runtime error.

**Comment by thautwarm** [comment]
Also, `Not` is something beyond my knowledge, as in other programming languages with traits, they don't have this. It seems very fancy, but I think I can come up with use cases for it. I'd ask people why there's not a `Not{X}` trait for indicating some specific trait is not implemented.

**Comment by thautwarm** [comment]
@McCoyBecker 

> In Rust for example, a trait bound like the above is compiled away - at compile time, Rust creates methods for dispatch for all types which satisfy the bound. But here, the bound is checked dynamically and could only throw a runtime error.

Of course, correct. But it doesn't mean the dispatch must be able to made in compiled time(but must check it validity). There're cases called **polymorphic recursive**, where the trait dispatch must be a little dynamic.

**Comment by McCoyBecker** [comment]
Ah okay - I think I see what you're saying but let me chew on that.

**Comment by AriMKatz** [comment]
> we cannot give constraints when defining a function.

Does this mean we cannot statically dispatch on a trait?

**Comment by thautwarm** [comment]
@AriMKatz No, in Julia, the compiler will know the fact of unsatisfying the constraints, but it won't complain.
It leads to another question that I used to talk with @Roger-luo : should we have a mode with compiler complaining about type errors.

**Comment by AriMKatz** [comment]
Understood but I'm wondering more about code flow than type checking. Can I make two methods of a function `foo`,  each a different implementation for either  `Add1` and `Add2` and have the dispatch inlined?

**Comment by thautwarm** [comment]
@AriMKatz  I'm not sure what you're saying about "either Add1 and Add2", I guess you're talking about two implementation of `Add1` or `Add2`, like

```julia
 @trait Add1{T} begin
           add1 :: [T] => T
  end
@implement Add1{Int} begin
        add1(x::Int) = x + 1
 end
 @implement Add1{Tuple{Int, Int}} begin
         add1(x) = (x[1] + 1, x[2] + 1)
  end
```
So firstly, methods of traits defined in CanonicalTraits can get inlined:
```
julia> @code_llvm add1(1)

;  @ none within `add1'
; Function Attrs: uwtable
define i64 @julia_add1_17260(i64) #0 {
top:
; ┌ @ REPL[6]:2 within `##add1#405#3'
; │┌ @ int.jl:53 within `+'
    %1 = add i64 %0, 1
; └└
  ret i64 %1
}

julia> @code_llvm add1((1, 1))

;  @ none within `add1'
; Function Attrs: uwtable
define void @julia_add1_17229([2 x i64]* noalias nocapture sret, [2 x i64] addrspace(11)* nocapture nonnull readonly dereferenceable(16)) #0 {
top:
; ┌ @ REPL[10]:2 within `##add1#409#5'
; │┌ @ int.jl:53 within `+'
    %2 = bitcast [2 x i64] addrspace(11)* %1 to <2 x i64> addrspace(11)*
    %3 = load <2 x i64>, <2 x i64> addrspace(11)* %2, align 8
    %4 = add <2 x i64> %3, <i64 1, i64 1>
; └└
  %5 = bitcast [2 x i64]* %0 to <2 x i64>*
  store <2 x i64> %4, <2 x i64>* %5, align 8
  ret void
}
```

And when using them with classic julia methods, things are still good:

```julia
julia> nadd1(x::Int) = add1(x)
nadd1 (generic function with 2 methods)

julia> nadd1(x::Tuple{Int, Int}) = add1(x)
nadd1 (generic function with 3 methods)

julia> @code_llvm nadd1((1, 1))

;  @ REPL[22]:1 within `nadd1'
; Function Attrs: uwtable
define void @julia_nadd1_17269([2 x i64]* noalias nocapture sret, [2 x i64] addrspace(11)* nocapture nonnull readonly dereferenceable(16)) #0 {
top:
; ┌ @ none within `add1'
; │┌ @ REPL[10]:2 within `##add1#409#5'
; ││┌ @ int.jl:53 within `+'
     %2 = bitcast [2 x i64] addrspace(11)* %1 to <2 x i64> addrspace(11)*
     %3 = load <2 x i64>, <2 x i64> addrspace(11)* %2, align 8
     %4 = add <2 x i64> %3, <i64 1, i64 1>
; └└└
  %5 = bitcast [2 x i64]* %0 to <2 x i64>*
  store <2 x i64> %4, <2 x i64>* %5, align 8
  ret void
}

julia> @code_llvm nadd1(1)

;  @ REPL[21]:1 within `nadd1'
; Function Attrs: uwtable
define i64 @julia_nadd1_17270(i64) #0 {
top:
; ┌ @ none within `add1'
; │┌ @ REPL[6]:2 within `##add1#405#3'
; ││┌ @ int.jl:53 within `+'
     %1 = add i64 %0, 1
; └└└
  ret i64 %1
}
```

We can treat a julia function as a single-method traits.

Julia functions can have multiple methods, and a single-method traits can also have multiple implementations.

However, multiple-method traits help to organize codebase.

**Comment by AriMKatz** [comment]
I meant "add 1 or add 2", but look like I'm not so clear on the definition of traits as used in this package and haskell/scala. I'll do some reading and work through your comments.  Thanks!

**Comment by AriMKatz** [comment]
So, I want to be able to dispatch on any type that has an implementation of add1, and then just call that implementation in the method body.

Let's say there's also an add2

so it would be

```
@CTraitdispatch f(x ) where x implements add1 = add1(x) + foo(x)

@CTraitdispatch f(x) where x implements add2= add2(x) +bar(x)
```

that way I'm dispatching on behavior rather than the concrete type.

**Comment by thautwarm** [comment]
@AriMKatz Very sorry that I didn't realized what you expected. It's already supported if your method `f` is a trait method.

see: https://github.com/thautwarm/CanonicalTraits.jl/issues/20#issuecomment-990660369

**Comment by Roger-luo** [comment]
FYI I think with GPUCompiler we can have a fully static subset and complain the not inferrable types now. But I'm not sure what would be a good practical use case to motivate this
